### PR TITLE
Implemented demultiplexing for rate limiting broadcasts.

### PR DESCRIPTION
**Pull request by pompon0**
This is a proof of concept, which might evolve with further refactors of the near-network package. Still we definitely we need some form of rate limiting of the messages we send. Thanks to demultiplexing:
- broadcasting is rate limited (which should prevent flooding attack)
- rate limited broadcasts are aggregated and send together (which prevents queueing a lot of requests)
- message handlers execution flow is easy to follow (and test)

### CODE REVIEW

**Review by matklad — COMMENTED** [review]
So, I think this basically looks good as a synchronization primitives. The API is cute!

I do have some doubts about whether this is the right tool for the job though. I have two vague concerns here:

* Today peers are actors, and actor's state machine seems like a good fit to manage various throttle queues. What this PR does is it introduces a separate state machine, which exists separately from the actor. So rather than composing two state machines into a bigger state machine, we runt them concurrently. If we do want to use actors for peers (which I don't know if we do), I'd prefer a solution which works on the actor's event loop, and not on a separate one. So, something like this: 
  * receive message from the per
  * put it into a vec with backlog
  * schedule a message to be delivered to oneself when the backlog needs pruning 
* Looking at the call-site

    ```rust
                    let (new_data, err) = pms.accounts_data.clone().insert(msg.accounts_data).await;
                    // Broadcast any new data we have found.
                    if new_data.len() > 0 {
                        pms.broadcast_accounts_data(new_data).await;
                    }
    ```

    it seems that we do aggergation/batching twice? That is, we don't actually want to batch the calls, we'd rather want to just broadcast the results of the latest call?

File: chain/network/src/concurrency/demux.rs [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+                        None => closed = true,
+                    },
+                }
+                if !calls.is_empty() && tokens > 0 {
+                    tokens -= 1;
+                    // TODO(gprusak): as of now Demux (as a concurrency primitive) doesn't support
+                    // cancellation. Once we add cancellation support, this task could accept a context sum:
+                    // the sum is valid iff any context is valid.
+                    let calls = std::mem::take(&mut calls);
+                    tokio::spawn(async move {
```
[the remark this answers is not in the document]
Comment (pompon0) [reply chain/network/src/concurrency/demux.rs:140]: Yes, the order of execution is not guaranteed and there is no reason to guarantee anything except for eventual execution: the calls are expected to be executed by independent callers which have no incentive to synchronize with each other for other reason than to save resources.

File: chain/network/src/peer_manager/peer_manager_actor.rs [hunk 2]
```diff
@@ -156,9 +157,57 @@ pub(crate) struct PeerManagerState {
...
+                        Some(x) => x.timestamp < d.timestamp,
+                    } {
+                        sum.insert(k, d);
+                    }
+                }
+                self.broadcast_message(SendMessage {
+                    message: PeerMessage::SyncAccountsData(SyncAccountsData {
+                        incremental: true,
+                        requesting_full_sync: false,
+                        accounts_data: sum.into_iter().map(|v| v.1).collect(),
```
[the remark this answers is not in the document]
Comment (pompon0) [reply chain/network/src/peer_manager/peer_manager_actor.rs]: done

File: chain/network/src/peer_manager/peer_manager_actor.rs [hunk 3]
```diff
@@ -156,9 +157,57 @@ pub(crate) struct PeerManagerState {
...
+                let res = ds.iter().map(|_| ()).collect();
+                let mut sum = HashMap::<_, SignedAccountData>::new();
+                for d in ds.into_iter().flatten() {
+                    let k = (d.epoch_id.clone(), d.account_id.clone());
+                    if match sum.get(&k) {
+                        None => true,
+                        Some(x) => x.timestamp < d.timestamp,
+                    } {
+                        sum.insert(k, d);
+                    }
```
[the remark this answers is not in the document]
Comment (pompon0) [reply chain/network/src/peer_manager/peer_manager_actor.rs]: done

**Comment by pompon0** [comment]
I've replaced per PeerManager batching with per Peer batching.

File: chain/network/src/network_protocol/mod.rs [hunk 4]
```diff
@@ -129,14 +130,15 @@ impl AccountKeySignedPayload {
...
     pub fn verify(&self, key: &PublicKey) -> bool {
         self.signature.verify(&self.payload, key)
     }
 }
 
-// TODO(gprusak): if we expect this to be large, perhaps we should
-// pass around an Arc, rather than pass it by value.
-#[derive(PartialEq, Eq, Clone, Debug, Hash)]
+#[derive(PartialEq, Eq, Debug, Hash)]
 pub struct SignedAccountData {
```
Comment (matklad) [inline chain/network/src/network_protocol/mod.rs:142]: This thing is effectively immutable (as it is signed), and we always pass an Arc of it. I think that this means that we probably should just put an `Arc` inside of it. Can be refactored separately from this PR.

Comment (pompon0) [reply chain/network/src/network_protocol/mod.rs:145]: added a TODO

File: chain/network/src/peer/peer_actor.rs [hunk 5]
```diff
@@ -1020,15 +1020,14 @@ impl StreamHandler<Result<Vec<u8>, ReasonForBan>> for PeerActor {
...
-                            requesting_full_sync: false,
-                            accounts_data: new_data,
-                        }))
-                        .await;
+                        let handles: Vec<_> = pms
+                            .connected_peers
+                            .read()
+                            .values()
+                            .map(|p| p.send_accounts_data(new_data.clone()))
+                            .collect();
```
Comment (matklad) [inline chain/network/src/peer/peer_actor.rs:1029]: Don't need `.collect` here?

File: chain/network/src/peer/peer_actor.rs [hunk 6]
```diff
@@ -1020,15 +1020,14 @@ impl StreamHandler<Result<Vec<u8>, ReasonForBan>> for PeerActor {
...
+                        futures::future::join_all(handles).await;
```
Comment (matklad) [inline chain/network/src/peer/peer_actor.rs:1030]: not-strong-opinion drive-by opinion: we shouldn't use the `futures` crate, we should use only `futures_util`.

File: chain/network/src/peer_manager/connected_peers.rs [hunk 7]
```diff
@@ -29,6 +35,49 @@ pub(crate) struct ConnectedPeer {
...
     /// Encoding used for communication.
     pub encoding: Option<Encoding>,
+    pub send_accounts_data_demux: demux::Demux<Vec<Arc<SignedAccountData>>, ()>,
+}
+
+impl ConnectedPeer {
+    pub fn send_accounts_data(
+        &self,
+        data: Vec<Arc<SignedAccountData>>,
+    ) -> impl std::future::Future<Output = ()> {
```
Comment (matklad) [inline chain/network/src/peer_manager/connected_peers.rs:45]: I'd maybe `use std::future::Future` -- future feels like a vocabulary typed designed for unqalified usage. Super weak opinion.

File: chain/network/src/peer_manager/connected_peers.rs [hunk 8]
```diff
@@ -29,6 +35,49 @@ pub(crate) struct ConnectedPeer {
...
+                    .send(SendMessage {
+                        message: PeerMessage::SyncAccountsData(SyncAccountsData {
+                            incremental: true,
+                            requesting_full_sync: false,
+                            accounts_data: sum.into_values().collect(),
+                        }),
+                        context: Span::current().context(),
+                    })
+                    .await
+                {
```
Comment (matklad) [inline chain/network/src/peer_manager/connected_peers.rs:74]: Maybe this wants `let msg =`? Struct literal in an `if` looks ugly.

File: chain/network/src/peer_manager/connected_peers.rs [hunk 9]
```diff
@@ -29,6 +35,49 @@ pub(crate) struct ConnectedPeer {
...
+                    error!("Failed sending incremental SyncAccountsData: {err}");
```
Comment (matklad) [inline chain/network/src/peer_manager/connected_peers.rs:75]:
```suggestion
                    tracing::error!(target: "network", ?err, "Failed sending incremental SyncAccountsData");
```


`?err` is a bit more structured (everything uses this specific name for errors). 

`"target"` is how all the code works. I wish we relied on default targets (module name), but not quite ready to fight that battle yet:)

Oh, and `tracing::` is just I think a little bit nicer, but no strong opinion.

**Review by matklad — APPROVED** [review]
This looks basically OK to me. I want to take a closer look tomorrow, but approving to unblock. 

Though, I would love to see some kind of stress-test here, to sanity-check that propagation of the info through the network doesn't require too many messages.

File: chain/network/src/peer/peer_actor.rs [hunk 10]
```diff
@@ -1020,15 +1020,14 @@ impl StreamHandler<Result<Vec<u8>, ReasonForBan>> for PeerActor {
...
-                    // unless there was no recent broadcast of this type.
                     if new_data.len() > 0 {
-                        pms.broadcast_message(PeerMessage::SyncAccountsData(SyncAccountsData {
-                            incremental: true,
-                            requesting_full_sync: false,
-                            accounts_data: new_data,
-                        }))
-                        .await;
+                        let handles: Vec<_> = pms
+                            .connected_peers
```
Comment (matklad) [inline chain/network/src/peer/peer_actor.rs:1025]: Let's filter-out the peer itself. That is, don't echo the new data back.

File: chain/network/src/peer/peer_actor.rs (hunk 6 above)
Comment (pompon0) [reply chain/network/src/peer/peer_actor.rs:1030]: done

File: chain/network/src/peer_manager/connected_peers.rs (hunk 7 above)
Comment (pompon0) [reply chain/network/src/peer_manager/connected_peers.rs:45]: done

File: chain/network/src/peer_manager/connected_peers.rs (hunk 8 above)
Comment (pompon0) [reply chain/network/src/peer_manager/connected_peers.rs:74]: done

File: chain/network/src/peer_manager/connected_peers.rs (hunk 9 above)
Comment (pompon0) [reply chain/network/src/peer_manager/connected_peers.rs:75]: done, +1 for default targets

File: chain/network/src/concurrency/demux.rs [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+
+    pub fn new(rl: RateLimit) -> Demux<Arg, Res> {
+        let (send, mut recv): (Stream<Arg, Res>, _) = mpsc::unbounded_channel();
+        tokio::spawn(async move {
+            let mut calls = vec![];
+            let mut closed = false;
+            let mut tokens = rl.burst;
+            let mut next_token = None;
+            let interval = (time::Duration::SECOND / rl.qps).try_into().unwrap();
+            while !calls.is_empty() || !closed {
```
[the remark this answers is not in the document]
Comment (pompon0) [reply chain/network/src/concurrency/demux.rs:132]: done

File: chain/network/src/concurrency/demux.rs [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+        async move {
+            let (send, recv) = oneshot::channel();
+            // ok().unwrap(), because DemuxCall doesn't implement Debug.
+            stream.send(Call { arg, out: send, handler: f.wrap() }).ok().unwrap();
+            recv.await.unwrap()
+        }
+    }
+
+    pub fn new(rl: RateLimit) -> Demux<Arg, Res> {
+        let (send, mut recv): (Stream<Arg, Res>, _) = mpsc::unbounded_channel();
```
[the remark this answers is not in the document]
Comment (pompon0) [reply chain/network/src/concurrency/demux.rs:125]: true, added panic if invalid, added a TODO to enforce RateLimit validation earlier.

File: chain/network/src/concurrency/demux.rs [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+                        let mut handlers = vec![];
+                        for call in calls {
+                            args.push(call.arg);
+                            outs.push(call.out);
+                            handlers.push(call.handler);
+                        }
+                        // A fancy way of extracting first element and dropping everything else.
+                        // Just calling (handler[0])(args).await, would require handlers to
+                        // implement Sync for some reason.
+                        let handler = handlers.into_iter().next().unwrap();
```
[the remark this answers is not in the document]
Comment (pompon0) [reply chain/network/src/concurrency/demux.rs:173]: thanks, that was tricky.

File: chain/network/src/concurrency/demux.rs [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+                        let res = (handler)(args).await;
+                        if res.len() != outs.len() {
+                            panic!(
+                                "demux handler returned {} results, expected {}",
+                                res.len(),
+                                outs.len()
+                            );
+                        }
+                        for (res, out) in res.into_iter().zip(outs.into_iter()) {
```
[the remark this answers is not in the document]
Comment (pompon0) [reply chain/network/src/concurrency/demux.rs:182]: done

File: chain/network/src/concurrency/demux.rs [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+                            outs.push(call.out);
+                            handlers.push(call.handler);
+                        }
+                        // A fancy way of extracting first element and dropping everything else.
+                        // Just calling (handler[0])(args).await, would require handlers to
+                        // implement Sync for some reason.
+                        let handler = handlers.into_iter().next().unwrap();
+                        let res = (handler)(args).await;
+                        if res.len() != outs.len() {
+                            panic!(
```
[the remark this answers is not in the document]
Comment (pompon0) [reply chain/network/src/concurrency/demux.rs:176]: done

File: chain/network/src/network_protocol/mod.rs [hunk 16]
```diff
@@ -129,14 +130,18 @@ impl AccountKeySignedPayload {
     pub fn len(&self) -> usize {
         self.payload.len()
     }
+    pub fn signature(&self) -> &near_crypto::Signature {
+        &self.signature
+    }
     pub fn verify(&self, key: &PublicKey) -> bool {
```
[the remark this answers is not in the document]
Comment (pompon0) [reply chain/network/src/network_protocol/mod.rs:136]: done

File: chain/network/src/peer/peer_actor.rs (hunk 10 above)
Comment (pompon0) [reply chain/network/src/peer/peer_actor.rs:1026]: done

File: chain/network/src/peer_manager/tests.rs [hunk 17]
```diff
@@ -306,3 +304,89 @@ async fn accounts_data_gradual_epoch_change() {
...
 }
+
+// Test is expected to take ~5s.
+// Test with 20 peer managers connected in layers:
+// - 1st 5 and 2nd 5 are connected in full bipartite graph.
+// - 2nd 5 and 3rd 5 ...
+// - 3rd 5 and 4th 5 ...
+// All of them are validators.
+#[tokio::test(flavor = "multi_thread")]
+async fn accounts_data_rate_limiting() {
```
Comment (matklad) [inline chain/network/src/peer_manager/tests.rs:315]: Great test!

**Pull request merged**